scripts/copper_bulk_prepare.py

```diff
@@ -0,0 +1,166 @@
+        "bulk_cu_nxnxn": f"bulk_cu_{k}x{k}x{k}",
+    }
+    calculation = Castep(
+        directory=directory,
+        label=label,
+        castep_command="castep.serial",
+        check_castep_version=True,
+        keyword_tolerance=0,
+    )
+    calculation.xc_functional = "PBE"
+    calculation.cut_off_energy = 340  # in eV
+    calculation.cell.kpoint_mp_grid = f"{k} {k} {k}"
+    # Controlling the angles within the cell so that it's FCC all the time during simulation
+    calculation.cell.cell_constraints = "1 1 1\n0 0 0"
+    calculation.task = "GeometryOptimization"
+
+    calculation._track_output = False  # type: ignore : bad library type
+    calculation._try_reuse = True
+    calculation._pedantic = True
+    calculation._rename_existing_dir = False
+    calculation.param.reuse = True
+    calculation.param.num_dump_cycles = 0
+    calculation._export_settings = True
+    # Attach the calculator to the atoms object
+    calculation.set_atoms(bulk_copper)
+    bulk_copper.set_calculator(calculation)
+
+    if calculation.dryrun_ok():
+        print("ok")
+
+        calculation.prepare_input_files()
+        # Write the submit script
+
+        with open(submit_script_from_path) as file:
+            content = file.read()
+
+        for n, k in replacements.items():
+            content = content.replace(n, k)
+
+        with open(submit_script_to_path, "w") as f:
+            f.write(content)
+    else:
+        msg = "dryrun failed"
+        raise Exception(msg)
+
+    # write a submit_all_script into data file
+    submit_all_script_from_path = "scripts/submit_all_script.sh"
+    submit_all_script_to_path = os.path.join("data", "submit_all.sh")
+    with open(submit_all_script_from_path) as file:
+        content = file.read()
+
+    with open(submit_all_script_to_path, "w") as f:
+        f.write(content)
+
+
+import os
+
+from ase.build import bulk
+from ase.calculators.castep import Castep
+
+
+def create_bulk_copper_structure():
+    """Create and return the bulk copper structure."""
+    return bulk("Cu", "fcc", a=3.8)
+
+
+def setup_calculation(directory, label, k):
+    """Set up the CASTEP calculation parameters."""
+    calculation = Castep(
+        directory=directory,
+        label=label,
+        castep_command="castep.serial",
+        check_castep_version=True,
+        keyword_tolerance=0,
+    )
+    calculation.xc_functional = "PBE"
+    calculation.cut_off_energy = 340  # in eV
+    calculation.cell.kpoint_mp_grid = f"{k} {k} {k}"
+    calculation.cell.cell_constraints = "1 1 1\n0 0 0"
+    calculation.task = "GeometryOptimization"
+
+    calculation._track_output = False  # type: ignore : bad library type
+    calculation._try_reuse = True
+    calculation._pedantic = True
+    calculation._rename_existing_dir = False
+    calculation.param.reuse = True
+    calculation.param.num_dump_cycles = 0
+    calculation._export_settings = True
+
+    return calculation
+
+
+def write_submit_script(
```

Comment:
i think this function should take job_name as an argument - the user should not have to figure out how to provide the correct replacements.

The template file could probably just live in the python source code rather than loading from a file

You should also add a get_castep_submit_script() which just creates a file, which is used by this function to create and save a file